### PR TITLE
Pre-compile localized strings and remove intl-messageformat

### PR DESCRIPTION
**Pull request by devongovett**
This reduces our bundle size by ~40 KB minified, or around ~11 KB min+gzip by removing the intl-messageformat dependency. This is done by pre-compiling messages during the build rather than parsing them at runtime. Basically, a message like this:

```json
"{count, plural, =0 {No items selected} one {# item selected} other {# items selected}}."
```

is compiled to a function like this, which returns a template literal:

```js
(args, formatter) => `${formatter.plural(args.count, {"=0": `No items selected`, one: () => `${formatter.number(args.count)} item selected`, other: () => `${formatter.number(args.count) items selected}`})}.`
```

As you can see, the `MessageFormatter` class includes some helpers used by the messages, such as plural formatting.

This example is kinda advanced, but in many cases, there are only basic interpolations without fancy formatting.

```js
(args) => `${args.item} selected`
```

And if there are no interpolations at all, then only a string is returned rather than a function.

This is a breaking change to `@internationalized/message` and therefore `@react-aria/i18n` since the input message format changed. We will need to decide how to handle this, especially with the mono package.

### CODE REVIEW

File: packages/@react-aria/dnd/intl/pt-PT.json [hunk 1]
```diff
@@ -3,7 +3,7 @@
   "dragDescriptionTouch": "Faça duplo toque para começar a arrastar.",
   "dragDescriptionVirtual": "Clique para iniciar o arrasto.",
   "dragItem": "Arrastar {itemText}",
-  "dragSelectedItems": "Arrastar {count, plural, one {# item selecionado} outro {# itens selecionados}}",
+  "dragSelectedItems": "Arrastar {count, plural, one {# item selecionado} other {# itens selecionados}}",
```
Comment (snowystinger) [inline packages/@react-aria/dnd/intl/pt-PT.json:6]: i think `outro` is correct here? https://translate.google.com/?sl=pt&tl=en&text=outro&op=translate&hl=en

Comment (devongovett) [reply packages/@react-aria/dnd/intl/pt-PT.json:6]: No, this should not have been translated. It's a keyword, not part of the message.

**Comment by devongovett** [comment]
Discussed with @dannify about how to handle the breaking change. We would like to avoid bumping `react-aria` to v4 until we have a few more breaking changes ready. But we also don't want to block this until then. Instead, we propose the following:

* Keep the existing `useMessageFormatter` hook for backward compatibility, but deprecate it, and remove from the sidebar in the docs.
* Introduce a new `useLocalizedStringFormatter` hook for the new version.
* Create a new `@internationalized/string` package and deprecate `@internationalized/message`. This would contain `LocalizedStringFormatter` and `LocalizedStringDictionary` objects to replace the existing `MessageFormatter` and `MessageDictionary`. This way the names of these objects match the hook.

Alternative would be to only bump a major on `@internationalized/message`, but then we'd need to move the old implementation into `@react-aria/i18n` in order to keep `useMessageFormatter` around. Plus, it might be nice to have the names match across the board.

File: packages/@react-aria/i18n/src/useLocalizedStringFormatter.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+}
+
+/**
+ * Provides localized string formatting for the current locale. Supports interpolating variables,
+ * selecting the correct pluralization, and formatting numbers. Automatically updates when the locale changes.
+ * @param strings - A mapping of languages to localized strings by key.
+ */
+export function useLocalizedStringFormatter<K extends string = string, T extends LocalizedString = string>(strings: LocalizedStrings<K, T>): LocalizedStringFormatter<K, T> {
+  let {locale} = useLocale();
+  let dictionary = useMemo(() => getCachedDictionary(strings), [strings]);
```
Comment (snowystinger) [inline packages/@react-aria/i18n/src/useLocalizedStringFormatter.ts:35]: won't useMemo already cache the dictionary?
```
useMemo(() => new LocalizedStringDictionary(strings), [strings]);
```

EDIT: this caches across components, useMemo will only cache within the component? in which case should we bother with the useMemo? seems like extra memory overhead

**Review by snowystinger — DISMISSED** [review]

File: packages/@react-aria/i18n/src/useLocalizedStringFormatter.ts (hunk 2 above)
Comment (snowystinger) [reply packages/@react-aria/i18n/src/useLocalizedStringFormatter.ts:35]: we did it this way before, discarding worry about memory for now

**Review by LFDanLu — APPROVED** [review]
Heads up, Field had an instance of useMessageFormatter that was added recently so I went ahead and updated that. Other than that, I assume we'll be updating `internationalized.mdx` to point to `useLocalizedStringFormatter` and adding a deprecation notice to `useMessageFormatter.mdx` after this goes in?

**Review by snowystinger — DISMISSED** [review]

**Comment by devongovett** [comment]
Yes, will update the docs separately.

**Pull request merged**